pyOCD/coresight/cortex_m.py

```diff
@@ -335,66 +284,47 @@ def __init__(self, name, bitsize, reg_type, reg_group):
-        self.breakpoints = {}
-        self.nb_code = 0
-        self.nb_lit = 0
-        self.num_hw_breakpoint_used = 0
-        self.nb_lit = 0
-        self.fpb_enabled = False
-        self.watchpoints = []
-        self.watchpoint_used = 0
-        self.dwt_configured = False
         self.arch = 0
         self.core_type = 0
         self.has_fpu = False
-        self.part_number = self.__class__.__name__
-        self.dap = Dap(link)
+        self.dp = dp
+        self.ap = ap
+        self.core_number = core_num
+
+        # Set up breakpoints manager.
+        self.fpb = FPB(self.ap)
+        self.dwt = DWT(self.ap)
+        self.sw_bp = SoftwareBreakpointProvider(self)
+        self.bp_manager = BreakpointManager(self)
+        self.bp_manager.add_provider(self.fpb, Target.BREAKPOINT_HW)
+        self.bp_manager.add_provider(self.sw_bp, Target.BREAKPOINT_SW)
 
     def init(self, initial_setup=True, bus_accessible=True):
         """
         Cortex M initialization
         """
-        if initial_setup:
-            self.link.connect()
-            self.idcode = self.readIDCode()
-            self.dap.init()
-
-            # select bank 0 (to access DRW and TAR)
-            self.dap.writeDP(DP_REG['SELECT'], 0)
-            self.dap.writeDP(DP_REG['CTRL_STAT'], CortexM.CSYSPWRUPREQ | CortexM.CDBGPWRUPREQ)
-
-            while True:
-                r = self.dap.readDP(DP_REG['CTRL_STAT'])
-                if (r & (CortexM.CDBGPWRUPACK | CortexM.CSYSPWRUPACK)) == (CortexM.CDBGPWRUPACK | CortexM.CSYSPWRUPACK):
-                    break
-
-            self.dap.writeDP(DP_REG['CTRL_STAT'], CortexM.CSYSPWRUPREQ | CortexM.CDBGPWRUPREQ | CortexM.TRNNORMAL | CortexM.MASKLANE)
-            self.dap.writeDP(DP_REG['SELECT'], 0)
-
-            ahb_idr = self.dap.readAP(AP_REG['IDR'])
-            if ahb_idr in AHB_IDR_TO_WRAP_SIZE:
-                self.auto_increment_page_size = AHB_IDR_TO_WRAP_SIZE[ahb_idr]
-            else:
-                # If unknown use the smallest size supported by all targets.
-                # A size smaller than the supported size will decrease performance
-                # due to the extra address writes, but will not create any
-                # read/write errors.
-                auto_increment_page_size = 0x400
-                logging.warning("Unknown AHB IDR: 0x%x" % ahb_idr)
-
         if bus_accessible:
             if self.halt_on_connect:
                 self.halt()
-            self.setupFPB()
             self.readCoreType()
             self.checkForFPU()
-            self.setupDWT()
             self.buildTargetXML()
+            self.fpb.init()
+            self.dwt.init()
+            self.sw_bp.init()
+
+    def disconnect(self):
+        # Remove breakpoints.
+        self.bp_manager.remove_all_breakpoints()
+
+        # Disable core debug.
+        self.write32(CortexM.DHCSR, CortexM.DBGKEY | 0x0000)
+
+        # Disable other debug blocks.
+        self.write32(CortexM.DEMCR, 0)
```

Comment:
Is there any reason the DEMCR is disabled before the DHCSR disabled debugging?  It might be safer to do in the reverse order.